File: solutions/elixir/resistor-color-trio/1/lib/resistor_color_trio.ex
defmodule ResistorColorTrio do
  @doc """
  Calculate the resistance value in ohms from resistor colors
  """
  @spec label(colors :: [atom]) :: {number, :ohms | :kiloohms | :megaohms | :gigaohms}
  def label(colors) do
    [first_color, second_color, multiplier_color | _] = colors

    first_digit = color_to_number(first_color)
    second_digit = color_to_number(second_color)
    multiplier = color_to_number(multiplier_color)

    base_value = first_digit * 10 + second_digit

    total_ohms = base_value * :math.pow(10, multiplier)

    format_resistance(trunc(total_ohms))
  end

  defp color_to_number(color) do
    case color do
      :black -> 0
      :brown -> 1
      :red -> 2
      :orange -> 3
      :yellow -> 4
      :green -> 5
      :blue -> 6
      :violet -> 7
      :grey -> 8
      :white -> 9
    end
  end

  defp format_resistance(ohms) when ohms < 1000 do
    {ohms, :ohms}
  end

  defp format_resistance(ohms) when ohms < 1_000_000 do
    {div(ohms, 1000), :kiloohms}
  end

  defp format_resistance(ohms) when ohms < 1_000_000_000 do
    {div(ohms, 1_000_000), :megaohms}
  end

  defp format_resistance(ohms) do
    {div(ohms, 1_000_000_000), :gigaohms}
  end
end

File: solutions/ruby/resistor-color-trio/1/resistor_color_trio.rb
class ResistorColorTrio
  COLORS_VALUES = %w[
    black
    brown
    red
    orange
    yellow
    green
    blue
    violet
    grey
    white
  ].freeze

  UNITS = %w[ohms kiloohms megaohms gigaohms].freeze

  attr_reader :colors

  def initialize(colors)
    @colors = colors
  end

  def label
    raise ArgumentError unless colors.all? { |color| COLORS_VALUES.include?(color) }

    resistance = first_two_colors * 10**third_color
    unit_index = 0

    while resistance >= 1000 && (resistance % 1000).zero? && unit_index < UNITS.length - 1
      resistance /= 1000
      unit_index += 1
    end

    "Resistor value: #{resistance} #{UNITS[unit_index]}"
  end

  def first_two_colors
    colors.take(2).map do |color|
      value(color)
    end.join.to_i
  end

  def third_color
    value(colors[2])
  end

  def value(index)
    COLORS_VALUES.index(index)
  end
end
